ycmd does not pass on many fixits

I'm finding that ycmd occasionally suggests fixits for things like missing semi-colons, but for many other things it does not suggest a fixit even though clang does. For instance, if I write `std::strng x;`, when compiling in clang it offers the correction to `string`, but ycmd does not offer this fixit. Is this a bug with ycmd, something wrong with my setup, or is the fixit feature constrained somehow?

In fact, it doesn't look like typo correction is kicking in in the following test case:

```c++
#include <string>

int main( int argc, char ** argv )
{
  std::strng this_is_a_string;
}
```

Diags reported:

```
fixit.cc|5 col 8 error| no type named 'strng' in namespace 'std'
fixit.cc|3 col 15 warning| unused parameter 'argc'
fixit.cc|3 col 29 warning| unused parameter 'argv'
```

And `:YcmShowDetailedDiagnostic`:

```
Forcing compilation, this will block Vim until done.
/Users/ben/Development/tmp/fixit.cc:5:8: error: no type named 'strng' in namespace 'std'
```

So, `libclang` isn't reporting a note or a FixIt (therefore there isn't much we can do).

Just to be sure, I checked and it is not related to the `Incomplete` flag that we use.

Maybe @r4nt or @d0k know if libclang would be expected to offer fixits for typo correction as in the above example? 

Compiling with clang `-fsyntax-only` gives this:

```
BeniMac:tmp ben$ clang -x c++ -fsyntax-only -Wall -Wextra -Werror fixit.cc
fixit.cc:5:8: error: no type named 'strng' in namespace 'std'; did you mean 'string'?
  std::strng this_is_a_string;
  ~~~~~^~~~~
       string
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/iosfwd:194:65: note: 'string' declared here
typedef basic_string<char, char_traits<char>, allocator<char> > string;
                                                                ^
1 error generated.
```

--- a/ycmd/completers/cpp/flags.py
+++ b/ycmd/completers/cpp/flags.py
@@ -99,7 +99,10 @@ class Flags( object ):
 
       if add_extra_clang_flags:
         flags += self.extra_clang_flags
-      sanitized_flags = PrepareFlagsForClang( flags, filename )
+
+      sanitized_flags = PrepareFlagsForClang( flags,
+                                              filename,
+                                              add_extra_clang_flags )
 
       if results[ 'do_cache' ]:
         self.flags_for_file[ filename ] = sanitized_flags
@@ -172,10 +175,12 @@ def _CallExtraConfFlagsForFile( module, filename, client_data ):
     return module.FlagsForFile( filename )
 
 
-def PrepareFlagsForClang( flags, filename ):
+def PrepareFlagsForClang( flags, filename, add_extra_clang_flags = True ):
   flags = _CompilerToLanguageFlag( flags )
   flags = _RemoveXclangFlags( flags )
   flags = _RemoveUnusedFlags( flags, filename )
+  if add_extra_clang_flags:
+    flags = _EnableTypoCorrection( flags )
   flags = _SanitizeFlags( flags )
   return flags
 
@@ -398,6 +403,24 @@ def _ExtraClangFlags():
   # for a similar issue.
   if OnWindows():
     flags.append( '-fno-delayed-template-parsing' )
+  return flags
+
+
+def _EnableTypoCorrection( flags ):
+  """Adds the -fspell-checking flag if the -fno-spell-checking flag is not
+  present"""
+
+  # "Typo correction" (aka spell checking) in clang allows it to produce
+  # hints (in the form of fix-its) in the case of certain diagnostics. A common
+  # example is "no type named 'strng' in namespace 'std'; Did you mean
+  # 'string'? (FixIt)". This is enabled by default in the clang driver (i.e. the
+  # 'clang' binary), but is not when using libclang (as we do). It's a useful
+  # enough feature that we just always turn it on unless the user explicitly
+  # turned it off in their flags (with -fno-spell-checking).
+  if '-fno-spell-checking' in flags:
+    return flags
+
+  flags.append( '-fspell-checking' )
   return flags
 
 

--- a/ycmd/tests/clang/flags_test.py
+++ b/ycmd/tests/clang/flags_test.py
@@ -185,6 +185,25 @@ def RemoveUnusedFlags_Depfiles_test():
                contains( *expected ) )
 
 
+def EnableTypoCorrection_Empty_test():
+  eq_( flags._EnableTypoCorrection( [] ), [ '-fspell-checking' ] )
+
+
+def EnableTypoCorrection_Trivial_test():
+  eq_( flags._EnableTypoCorrection( [ '-x', 'c++' ] ),
+                                    [ '-x', 'c++', '-fspell-checking' ] )
+
+
+def EnableTypoCorrection_Reciprocal_test():
+  eq_( flags._EnableTypoCorrection( [ '-fno-spell-checking' ] ),
+                                    [ '-fno-spell-checking' ] )
+
+
+def EnableTypoCorrection_ReciprocalOthers_test():
+  eq_( flags._EnableTypoCorrection( [ '-x', 'c++', '-fno-spell-checking' ] ),
+                                    [ '-x', 'c++', '-fno-spell-checking' ] )
+
+
 def RemoveUnusedFlags_RemoveFilenameWithoutPrecedingInclude_test():
   def tester( flag ):
     expected = [ 'clang', flag, '/foo/bar', '-isystem/zoo/goo' ]

--- a/ycmd/tests/clang/subcommands_test.py
+++ b/ycmd/tests/clang/subcommands_test.py
@@ -727,12 +727,32 @@ def FixIt_Check_cpp11_Note( results ):
   } ) )
 
 
+def FixIt_Check_cpp11_SpellCheck( results ):
+  assert_that( results, has_entries( {
+    'fixits': contains(
+      # Change to SpellingIsNotMyStrongPoint
+      has_entries( {
+        'text': contains_string( "did you mean 'SpellingIsNotMyStrongPoint'" ),
+        'chunks': contains(
+          ChunkMatcher( 'SpellingIsNotMyStrongPoint',
+                        LineColMatcher( 72, 9 ),
+                        LineColMatcher( 72, 35 ) )
+        ),
+        'location': LineColMatcher( 72, 9 ),
+      } ) )
+  } ) )
+
+
 def Subcommands_FixIt_all_test():
   cfile = 'FixIt_Clang_cpp11.cpp'
   mfile = 'FixIt_Clang_objc.m'
   ufile = 'unicode.cc'
 
   tests = [
+    # L
+    # i   C
+    # n   o
+    # e   l   Lang     File,  Checker
     [ 16, 0,  'cpp11', cfile, FixIt_Check_cpp11_Ins ],
     [ 16, 1,  'cpp11', cfile, FixIt_Check_cpp11_Ins ],
     [ 16, 10, 'cpp11', cfile, FixIt_Check_cpp11_Ins ],
@@ -762,6 +782,9 @@ def Subcommands_FixIt_all_test():
 
     # FixIt attached to a "child" diagnostic (i.e. a Note)
     [ 60, 1,  'cpp11', cfile, FixIt_Check_cpp11_Note ],
+
+    # FixIt due to forced spell checking
+    [ 72, 9,  'cpp11', cfile, FixIt_Check_cpp11_SpellCheck ],
   ]
 
   for test in tests:

--- a/ycmd/tests/clang/testdata/FixIt_Clang_cpp11.cpp
+++ b/ycmd/tests/clang/testdata/FixIt_Clang_cpp11.cpp
@@ -63,3 +63,11 @@ void z() {
 
   }
 }
+
+namespace Typo {
+    struct SpellingIsNotMyStrongPoint;
+}
+
+void typo() {
+  Typo::SpellingIsNotMyStringPiont *p;
+}

